### ISSUE TITLE
Ability to specify several class file directories for report

### ISSUE THREAD

**Issue by amaembo** [jacoco/jacoco#965]
### Scenario

* JaCoCo version: 0.8.4.201905082037
* Tool integration: Maven
* Description of your use case: my build has several output directories (namely target/classes and target/classes-9) because I use [multirelease jar plugin](https://github.com/metlos/multi-release-jar-maven-plugin). When running coverage I want to test classfiles from both of them. Surefire plugin allows me to do this specifying custom classpath via `classesDirectory` and `additionalClasspathElements`. And looks like JaCoCo properly gathers the stats for `classes-9` classes (at least their names appear inside `jacoco.exec`. The only missing link is the `report`. It looks like the `report` target considers class files from `project.getBuild().getOutputDirectory()` only, and there's no way to specify several output directories.

### Current Behaviour
It's impossible to specify several output directories to be considered during the report generation.

### Wanted Behaviour
It should be possible to specify several class-file directories.

### Possible Workarounds
For source directories, I can tweak this with build-helper-maven-plugin:add-sources, but the output directory is still only one. I could create a custom step before report to copy all the class files from all the output directories into the separate directory and modify build output directory to point to that directory, but then I need to set it back. Otherwise, I may get a corrupted jar. I haven't tried yet, but it seems possible, though it requires much more ugly steps in pom and makes the whole thing more fragile.

**Comment by marchof** [comment]
@amaembo @Godin We need to consider that JaCoCo cannot handle different versions of the same class file in the same report (group). Typically multirelease JARs will have different versions of the same class file, correct?

**Comment by amaembo** [comment]
@marchof for me this is a minor problem because I decided to reduce this as much as possible. Other tools could be not very happy with this as well. My class layout is close to this:

```java
// common src:
interface VersionSpecific {
  // methods which implementation differs for different Java versions
}
class Java8Specific extends VersionSpecific { ... }
final class VerSpec {
  VersionSpecific VER_SPEC = new Java8Specific();
}

// java 9 src
class Java9Specific extends VersionSpecific { ... }
final class VerSpec {
  VersionSpecific VER_SPEC = new Java9Specific();
}
```

So I have the only name clash for very short `VerSpec` class which is always covered. So having either of them in coverage report is fine for me. I agree that for other projects this could be a problem.

**Comment by marchof** [comment]
@amaembo Even if there is a single name clash and two versions of a class with the same name will be provided JaCoCo report generation will fail.

**Comment by Godin** [comment]
As was correctly pointed by @marchof - Multi-Release JAR Files always have at least two class files with same fully qualified name, compiled from two different source files (http://openjdk.java.net/jeps/238 , https://docs.oracle.com/en/java/javase/11/docs/specs/jar/jar.html).

The only way for JaCoCo to find source file for class file - is to rely on fully qualified name, so within same "group" they must be unique for generation of proper report. There is no reliable ways to automatically split class and source files on groups by version. The only way is to rely on information provided by user or by tool into which JaCoCo integrates.

@amaembo as you correctly pointed - for single module Maven Core has only `project.getBuild().getOutputDirectory()` and so doesn't provide such information.

Thus to cover case of multiple class files with same fully qualified name within single Maven module, user of `jacoco-maven-plugin:report` will be forced to provide the same configuration as one that was used by me for `maven-antrun-plugin` to resolve your issue (https://github.com/amaembo/streamex/pull/203). 

In this case as a user I don't see benefits of having first one over last one, and as one of only two spare-time developers of JaCoCo don't see reasons to spend time on such implementation and its future maintenance (duplicated code, documentation, questions, etc).

Also should not be forgotten that many users use `jacoco:report-aggregate`, which also uses `project.getBuild().getOutputDirectory()` and so also requires absence of class and source files with identical qualified names within individual modules (https://github.com/jacoco/jacoco/blob/v0.8.5/jacoco-maven-plugin/src/org/jacoco/maven/ReportSupport.java). This one is already a workaround over non-flexible Maven lifecycle, raising a lot of questions and maintenance pains on its own without even thinking how to update it for this case (https://github.com/jacoco/jacoco/issues/869).

However BTW `jacoco-maven-plugin:report-aggregate` can already be used for the case when multiple Maven modules are used to construct multi-release JAR.

And BTW while `maven-compiler-plugin` has option `multiReleaseOutput` AFAIK Maven Core Team recommends (CC @rfscholte) to use multiple modules to build multi release JARs (https://maven.apache.org/plugins/maven-compiler-plugin/multirelease.html), whereas description of https://github.com/metlos/multi-release-jar-maven-plugin

> This plugin is a quick and **dirty solution** to the current lack of tooling.

All in all: given the above, our limited resources, and most importantly existence of two alternatives (`maven-antrun-plugin` and `jacoco:report-aggregate`) and possibility to implement such thing(s) as third-party Maven plugin, I propose to close this ticket.

@amaembo @marchof WDYT?

**Comment by marchof** [comment]
I agree with @Godin that we should not try to target this in JaCoCo for now.

**Comment by amaembo** [comment]
Well, as I already have a solution for now and it's not very ugly, I agree that the ticket could be closed. Thank you for the help.
